x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AbstractPreserveAllocations.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.assignment.planning;
+
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlan.Model;
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlan.Node;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+abstract class AbstractPreserveAllocations {
+
+    private final List<Node> nodes;
+    private final List<Model> models;
+
+    protected AbstractPreserveAllocations(List<Node> nodes, List<Model> models) {
+        this.nodes = Objects.requireNonNull(nodes);
+        this.models = Objects.requireNonNull(models);
+    }
+
+    List<Node> nodesPreservingAllocations() {
+        return nodes.stream().map(n -> modifyNodePreservingAllocations(n)).toList();
+    }
+
+    private Node modifyNodePreservingAllocations(Node n) {
+        long bytesUsed = 0;
+        int coresUsed = 0;
+        for (Model m : models) {
+            if (m.currentAllocationByNodeId().containsKey(n.id())) {
+                bytesUsed += m.memoryBytes();
+                coresUsed += calculateUsedCores(n, m);
+            }
+        }
+
+        return new Node(n.id(), n.availableMemoryBytes() - bytesUsed, n.cores() - coresUsed);
+    }
+
+    List<Model> modelsPreservingAllocations() {
+        return models.stream().map(m -> modifyModelPreservingPreviousAssignments(m)).toList();
+    }
+
+    Model modifyModelPreservingPreviousAssignments(Model m) {
+        if (m.currentAllocationByNodeId().isEmpty()) {
+            return m;
+        }
+
+        return new Model(
+            m.id(),
+            m.memoryBytes(),
+            m.allocations() - calculatePreservedAllocations(m),
+            m.threadsPerAllocation(),
+            calculateAllocationsPerNodeToPreserve(m)
+        );
+    }
+
+    AssignmentPlan mergePreservedAllocations(AssignmentPlan assignmentPlan) {
+        final Map<Tuple<String, String>, Integer> assignmentsByModelNodeIdPair = new HashMap<>();
+        for (Model m : assignmentPlan.models()) {
+            Map<Node, Integer> assignments = assignmentPlan.assignments(m);
+            if (assignments != null) {
+                for (Map.Entry<Node, Integer> nodeAssignment : assignments.entrySet()) {
+                    assignmentsByModelNodeIdPair.put(Tuple.tuple(m.id(), nodeAssignment.getKey().id()), nodeAssignment.getValue());
+                }
+            }
+        }
```

Comment:
The problem is that the model/node objects of the `assignmentPlan` here could differ to the model/node objects stored in `models` and `nodes` members. The reason is that we modified them in order to account for previous assignments before computing the plan. That is why I create the lookup table based on `String` ids instead of the model/node objects themselves. I'll go back and add a comment to clarify that.